### PR TITLE
pybind/mgr/autoscaler: don't scale pools with overlapping roots

### PR DESCRIPTION
**Pull request by kamoltat**
In the previous version of get_subtree_resource_status() in
src/pybind/mgr/pg_autoscaler/module.py we ignore overlapping
pools which in some cases if combined with the new `scale-down`
algorithm in https://github.com/ceph/ceph/pull/38805 can cause
some pools to scale up/down to inapproriate amount of pgs.

Therefore, the PR identifies the overlapping roots and prevent the pools
with such roots from scaling. This only happens with `scale-down` profile
as we see no problem with the default `scale-up` profile.

Removed the variable `pool_root` since it is not used anywhere in
the code, it only gets assigned and reassigned.

Also included a unit test test_overlapping_roots.py that tests the function
identify_subtrees_and_overlaps() as well as edited test_cal_final_pg_target.py
to account for pools that contain overlapping roots, therefore, those pools
are expected not to scale.

Signed-off-by: Kamoltat <ksirivad@redhat.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by kamoltat** [comment]
jenkins test api

**Comment by kamoltat** [comment]
jenkins test api

**Comment by kamoltat** [comment]
jenkins test api

File: src/pybind/mgr/pg_autoscaler/module.py [hunk 1]
```diff
@@ -641,9 +655,6 @@ def _maybe_adjust(self):
                 'count': len(too_many),
                 'detail': too_many
             }
-
-        too_much_target_bytes = []
-        for root_id, total in total_bytes.items():
```
Comment (jdurgin) [inline src/pybind/mgr/pg_autoscaler/module.py:646]: these lines look removed by accident

**Review by jdurgin — COMMENTED** [review]
logic looks good! could you add printing out the overlapped_roots in the error message, to make it easier to diagnose?

a few whitespace changes look unrelated, but not a big deal

**Review by jdurgin — APPROVED** [review]

**Comment by kamoltat** [comment]
https://pulpito.ceph.com/ksirivad-2021-07-11_01:45:00-rados-wip-pg-autoscaler-overlap-distro-basic-smithi/

**Comment by tchaikov** [comment]
> https://pulpito.ceph.com/ksirivad-2021-07-11_01:45:00-rados-wip-pg-autoscaler-overlap-distro-basic-smithi/

took a look at [6262857](https://pulpito.ceph.com/ksirivad-2021-07-11_01:45:00-rados-wip-pg-autoscaler-overlap-distro-basic-smithi/6262857/). it seems that it's already tracked by https://tracker.ceph.com/issues/50192 .

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2021-07-17_04:37:01-rados-wip-kefu-testing-2021-07-17-1006-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-07-17_15:53:41-rados-wip-kefu-testing-2021-07-17-1518-distro-basic-smithi/

one of the failures is tracked by:

- https://tracker.ceph.com/issues/50245

the e2e failures are not related.

**Pull request merged**